pylanetary/spectral/core.py

```diff
@@ -0,0 +1,808 @@
+    
+    Returns
+    -------
+    mask : bool
+          A 2D boolean array where pixels with values of True are extracted, and values of False are not.
+    
+    '''
+    
+    mask = np.full((self.xpixmax,self.ypixmax), 0)
+    for i in pixels:
+      ix = i[0]
+      iy = i[1]
+      mask[ix,iy] = 1
+    mask = np.array(mask,dtype='bool')
+      
+    return mask
+    
+  ###############################################################################  
+    
+  def extract_mask_region(self,mask):
+    '''
+    Extract pixels defined in a boolean array.
+    
+    Parameters
+    ----------
+    mask : bool
+          2D boolean array of the size [xpixmax,ypixmax]. Pixels with a value of True are extracted for fitting.
+    
+    Returns
+    -------
+    datas : numpy array
+          The 3D numpy array containing the spectral axis. Masked pixels are NAN. 
+    
+    '''
+    datas = np.empty((self.xpixmax,self.ypixmax),dtype=np.ndarray)
+    
+    for xpix in range (0,self.xpixmax,1):
+      for ypix in range (0,self.ypixmax,1):
+      
+        if mask[xpix,ypix] == True:
+          datas[xpix,ypix] = self.data[0,:,ypix,xpix]
+        else:
+          datas[xpix,ypix] = np.NAN
+          
+    return datas
+  
+  
+  ###############################################################################
+  
+  def fit_data(self, datas, fit_type, outfile, initial_fit_guess = None, RMS = None, SN = None):
+    '''
+    Calculate the best fit for the spectra at each pixel.
+    
+    Parameters
+    ----------
+    datas : numpy array
+          The array of spectral data to be fit.
+    fit_type : string
+          The type of line shape used to fit the spectra. Options are Gaussian, Lorentzian, Voigt, PseudoV, Moffat, and All. It is not recommended that you use the option of all for anything more than a single pixel.
+    outfile : string
+          The name of the saved output pickle file. The 'pickle' file extension is automatically included.
+    initial_fit_guess : string, optional
+          The name of a saved LMFIT result to use as an intitial guess instead of the LMFIT guess function. Default is None.
+    RMS : float, optional
+          The numerical value of the data's RMS. Default is None; if not provided, an estimate is calculated using Astropy's siga_clipped_stats (https://docs.astropy.org/en/stable/api/astropy.stats.sigma_clipped_stats.html)
+    SN : float
+          The Signal to noise ratio minimum, which is used to mask out noisy data. Any pixel with a S/N ratio lower than this calue will not be fit. Default is value is 1. 
+    
+    '''
+    
+    if SN == None:
+      SN = 1
+    
+    picklefile = open(outfile+'.pickle', 'wb')
+    print('Numerical results will be saved in the '+outfile+'.pickle file')
+    results = {'x':[],'y':[],'center':[],'std':[],'chi':[],'redchi':[]}
+    
+    for xpix in range (0,self.xpixmax,1):
```

Comment:
why does this happen over self.xpixmax and self.ypixmax, when you have already specified the input data? shouldn't it just be applied to all the data that were requested by this function itself?